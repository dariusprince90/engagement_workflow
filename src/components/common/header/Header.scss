@import '~@Company/web-theme-03/src/scss/variables';
@import '../../../scss/variables';

header.common-page-header {
  background-color: $pm-blue;
  height: 80px;
  overflow: hidden;

  // flex container properties
  display: flex;
  flex-direction: row;
  align-items: center;

  img.pm-logo {
    height: 50px;
    margin-left: 0.25rem;
  }

  .titles-container {
    margin-left: 1rem;
    color: $white;
    white-space: nowrap;

    // flex item properties
    flex-grow: 1;

    // flex container properties
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    justify-content: center;
  }

  .page-title {
    font-size: 1.6rem;
    font-weight: 500;
    margin-bottom: 0.375rem;
  }

  .page-subtitle {
    font: normal 0.8rem $font-family-sans-serif-001;
    margin-bottom: 0.5rem;
    margin-left: 0.0625rem;
  }

  .page-help-container {
    margin-right: 0.5rem;

    // flex item properties
    flex-grow: 0;

    a {
      text-decoration: none;

      .help-logo {
        color: $pm-yellow;
        margin-right: 0.25rem;
      }

      .page-help-text {
        color: $white;
      }
    }
  }
}
