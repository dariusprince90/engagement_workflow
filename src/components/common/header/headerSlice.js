import { createSlice } from '@reduxjs/toolkit';

// **********************************************************************
// * constants

// **********************************************************************
// * functions

// **********************************************************************
// * thunks

// **********************************************************************
// * initial state

const initialState = { pageTitle: '', pageSubtitle: '' };

// **********************************************************************
// * slice

const headerSlice = createSlice({
  name: 'pageHeader',
  initialState,

  reducers: {
    pageTitleChanged: (state, action) => {
      state.pageTitle = action.payload;
    },

    pageSubtitleChanged: (state, action) => {
      state.pageSubtitle = action.payload;
    }
  },

  extraReducers: (builder) => {}
});

// **********************************************************************
// * actions

export const { pageTitleChanged, pageSubtitleChanged } = headerSlice.actions;

// **********************************************************************
// * selectors

export const selectPageTitle = (state) => state.pageHeader.pageTitle;
export const selectPageSubtitle = (state) => state.pageHeader.pageSubtitle;

// **********************************************************************
// * reducer

export default headerSlice.reducer;
