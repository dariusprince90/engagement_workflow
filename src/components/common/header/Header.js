import { useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import './Header.scss';
import { selectPageTitle, selectPageSubtitle } from './headerSlice';

const Header = () => {
  // **********************************************************************
  // * constants

  const pageTitle = useSelector(selectPageTitle);
  const pageSubtitle = useSelector(selectPageSubtitle);

  // **********************************************************************
  // * component vars

  // **********************************************************************
  // * functions

  // **********************************************************************
  // * side effects

  // **********************************************************************
  // * render

  return (
    <header className="common-page-header">
      {/* pm logo */}
      <Link to="/" reloadDocument>
        <img src="/images/pm-logos/pm-logo-white-small.png" alt="Company logo" className="pm-logo" />
      </Link>

      <span className="titles-container">
        {/* page title and subtitle */}
        <h3 className="page-title">{pageTitle}</h3>
        <h5 className="page-subtitle">{pageSubtitle}</h5>
      </span>

      <div className="page-help-container">
        <a
          href={
            'https://Company.sharepoint.com/sites/kshare-Internal-Accounting/Shared%20Documents/Forms/AllItems.aspx' +
            '?id=%2Fsites%2Fkshare%2DInternal%2DAccounting%2FShared%20Documents%2FNEW%20Best%20Practices%20and' +
            '%20Frequently%20Asked%20Questions%2Epdf&parent=%2Fsites%2Fkshare%2DInternal%2DAccounting%2FShared%20Documents'
          }
          target="_blank"
          rel="noopener noreferrer">
          <FontAwesomeIcon icon={['fas', 'question-circle']} className="help-logo" />
          <span className="page-help-text">Best Practices and FAQs</span>
        </a>
      </div>
    </header>
  );
};

export default Header;
