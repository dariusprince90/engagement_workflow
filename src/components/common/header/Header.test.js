/* eslint-disable react/prop-types */

import ReactDOM from 'react-dom';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import faker from '@faker-js/faker';
import { MemoryRouter } from 'react-router-dom';

import * as headerSlice from './headerSlice';
import Header from './Header';

// **********************************************************************
// * constants

const pageTitle = faker.datatype.string();
const pageSubtitle = faker.datatype.string();
const pageTitleForSnapshot = 'page title for snapshot tests';
const pageSubtitleForSnapshot = 'page subtitle for snapshot tests';
const testIds = {
  fontAwesomeIcon: 'font-awesome-icon'
};

const componentToRender = (
  <MemoryRouter>
    <Header />
  </MemoryRouter>
);

// **********************************************************************
// * mock external dependencies

jest.mock('@fortawesome/react-fontawesome', () => {
  return {
    FontAwesomeIcon: ({ icon, className }) => {
      return <fake-font-awesome-icon icon={icon} className={className} data-testid={testIds.fontAwesomeIcon} />;
    }
  };
});

jest.mock('react-redux', () => {
  return {
    useSelector: (callback) => {
      return callback();
    }
  };
});

jest.mock('./headerSlice');

// **********************************************************************
// * unit tests

describe('Header', () => {
  // **********************************************************************
  // * setup

  beforeEach(() => {
    headerSlice.selectPageTitle.mockReturnValue(pageTitle);
    headerSlice.selectPageSubtitle.mockReturnValue(pageSubtitle);
  });

  // **********************************************************************
  // * tear-down

  // **********************************************************************
  // * execution

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(componentToRender, div);
  });

  it('renders correctly and matches snapshot', () => {
    headerSlice.selectPageTitle.mockReturnValue(pageTitleForSnapshot);
    headerSlice.selectPageSubtitle.mockReturnValue(pageSubtitleForSnapshot);
    const tree = renderer.create(componentToRender).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the pm logo', () => {
    render(componentToRender);
    expect(screen.getByAltText('Company logo')).toBeInTheDocument();
  });

  it('renders the help icon', () => {
    render(componentToRender);
    expect(screen.getByTestId(testIds.fontAwesomeIcon)).toHaveAttribute('icon', 'fas,question-circle');
  });

  it('renders correct Help text', () => {
    const helpText = 'Best Practices and FAQs';
    render(componentToRender);
    expect(screen.getByText(helpText)).toBeInTheDocument();
  });

  it('renders correct Help link', () => {
    const helpText = 'Best Practices and FAQs';
    render(componentToRender);
    expect(screen.getByRole('link', { name: helpText })).toHaveAttribute(
      'href',
      'https://Company.sharepoint.com/sites/kshare-Internal-Accounting/Shared%20Documents/Forms/AllItems.aspx' +
        '?id=%2Fsites%2Fkshare%2DInternal%2DAccounting%2FShared%20Documents%2FNEW%20Best%20Practices%20and' +
        '%20Frequently%20Asked%20Questions%2Epdf&parent=%2Fsites%2Fkshare%2DInternal%2DAccounting%2FShared%20Documents'
    );
  });
});
