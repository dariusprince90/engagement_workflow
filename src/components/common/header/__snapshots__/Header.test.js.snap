// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly and matches snapshot 1`] = `
<header
  className="common-page-header"
>
  <a
    href="/"
  >
    <img
      alt="Company logo"
      className="pm-logo"
      src="/images/pm-logos/pm-logo-white-small.png"
    />
  </a>
  <span
    className="titles-container"
  >
    <h3
      className="page-title"
    >
      page title for snapshot tests
    </h3>
    <h5
      className="page-subtitle"
    >
      page subtitle for snapshot tests
    </h5>
  </span>
  <div
    className="page-help-container"
  >
    <a
      href="https://Company.sharepoint.com/sites/kshare-Internal-Accounting/Shared%20Documents/Forms/AllItems.aspx?id=%2Fsites%2Fkshare%2DInternal%2DAccounting%2FShared%20Documents%2FNEW%20Best%20Practices%20and%20Frequently%20Asked%20Questions%2Epdf&parent=%2Fsites%2Fkshare%2DInternal%2DAccounting%2FShared%20Documents"
      rel="noopener noreferrer"
      target="_blank"
    >
      <fake-font-awesome-icon
        className="help-logo"
        data-testid="font-awesome-icon"
        icon={
          Array [
            "fas",
            "question-circle",
          ]
        }
      />
      <span
        className="page-help-text"
      >
        Best Practices and FAQs
      </span>
    </a>
  </div>
</header>
`;
